presto-jdbc/src/main/java/io/prestosql/jdbc/AbstractPrestoResultSet.java

```diff
@@ -1788,11 +1803,25 @@ private static Timestamp parseTimestamp(String value, Function<String, ZoneId> t
                 .atZone(zoneId)
                 .toEpochSecond();
 
-        Timestamp timestamp = new Timestamp(epochSecond * 1000);
+        Timestamp timestamp;
+        if (modernDate(year)) {
+            timestamp = new Timestamp(epochSecond * 1000);
+        }
+        else {
+            GregorianCalendar calendar = new GregorianCalendar(year, month - 1, day, hour, minute, second);
+            calendar.setTimeZone(TimeZone.getTimeZone(zoneId));
+            timestamp = new Timestamp(calendar.getTimeInMillis());
+        }
         timestamp.setNanos((int) rescale(fractionValue, precision, 9));
         return timestamp;
     }
 
+    private static boolean modernDate(int year)
+    {
+        // before this year, the chronology used by Joda is not consistent with java.sql.Date
+        return year >= 1900;
```

Comment:
Introduce constants

```
// Before that year, Java Time and Joda Time are not consistent with java.sql.Date and java.util.Calendar
// Since January 1, 1900 in one zone is still December 31, 1899 in some other zone, we are adding 1 year margin.
MODERN_YEAR = 1901

MODERN_YEAR_MILLIS_UTC = LocalDate.of(MODERN_YEAR,1,1)...........
```

Then you can inline modernDate